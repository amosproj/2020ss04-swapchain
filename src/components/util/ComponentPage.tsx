import React from "react"

interface Props {
  title?: string
  children: any
  bg?: string
}

export const ComponentPage = (props: Props) => {
  return (
    <div className="container mx-auto">
      <main className={`pb-10 mx-auto ${props.bg} ${props.bg ? "shadow-sm" : ""}  p-4 sm:px-6 lg:px-8`}>
        <h1 className="py-4 text-3xl font-semibold leading-tight text-gray-900">{props.title}</h1>
        {props.children}
      </main>
    </div>
  )
}
