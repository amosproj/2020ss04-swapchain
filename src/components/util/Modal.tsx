import React from "react"
import { ReactComponent as X } from "../../icons/x.svg"
type Props = {
  close: () => void
  open: boolean
  children: any
  title?: string
}

export const Modal = (props: Props) => {
  return (
    <div className={props.open ? "" : "hidden"}>
      <div className="absolute inset-0 bg-gray-500 opacity-75"></div>
      <div className="absolute inset-0 z-10 flex items-center justify-center">
        <div>
          <header className="flex items-center justify-between">
            <h1 className="text-lg font-semibold text-gray-500">{props.title}</h1>
            <X className="w-8 h-8 text-gray-400 duration-200 ease-in-out hover:text-gray-800" onClick={props.close}></X>
          </header>
          <div className="flex flex-col items-center justify-around flex-grow md:flex-row">{props.children}</div>
        </div>
      </div>
    </div>
  )
}
