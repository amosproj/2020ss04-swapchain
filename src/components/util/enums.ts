export enum State {
  IDLE,
  RUNNING,
  SUCCESS,
  ERROR,
  FAILURE,
}

export enum Currency {
  BTC = "BTC",
  BTS = "BTS",
}

export enum Network {
  MAINNET = "mainnet",
  TESTNET = "testnet",
}

export enum Priority {
  HIGH,
  MEDIUM,
  LOW,
}
