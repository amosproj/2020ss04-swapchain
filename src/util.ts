import * as bitcoin from "bitcoinjs-lib"

export const fakeKey = (length: number, network?: string) => {
  let prefix = ""
  if (network === "mainnet") {
    prefix = "L"
  } else if (network === "testnet") {
    prefix = "c"
  }

  return prefix + hash(Math.random().toString()).substring(0, length)
}

export const hash = (s: string): string => {
  return bitcoin.crypto.hash256(Buffer.from(s)).toString("hex")
}

export const toPublicKey = (privateKey: string): string => {
  const keyPair = bitcoin.ECPair.fromWIF(privateKey, bitcoin.networks.testnet)
  const keyPairCompressed = bitcoin.ECPair.fromPrivateKey(keyPair.privateKey!, { compressed: true })
  return keyPairCompressed.publicKey.toString("hex")
}
