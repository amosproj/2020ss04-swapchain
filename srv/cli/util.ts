import readline from "readline"

/**
 * Get input from stdin
 *
 * @param str - A question to ask the user.
 * @returns answer from the user
 */
export async function askUser(str: string): Promise<string> {
  const read = readline.createInterface({
    input: process.stdin,
    output: process.stdout,
  })

  return new Promise((resolve) => read.question("\u001b[36;1m" + str + "\u001b[0m", resolve))
}
