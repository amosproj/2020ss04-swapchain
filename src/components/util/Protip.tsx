import React from "react"
import { ReactComponent as LightBulb } from "../../icons/light-bulb.svg"
import { Network } from "../util/enums"

interface Props {
  network: Network
}

export const Protip = (props: Props) => {
  return (
    <div className="flex flex-col items-center px-3 py-2 space-x-1 md:flex-row">
      <div className="flex items-center">
        <LightBulb className="h-6 text-gray-600"></LightBulb>
        <span className="font-semibold text-gray-900">ProTip!</span>
      </div>
      <div className="flex flex-col items-center md:flex-row">
        <span className="text-center text-gray-700">Check your transaction on the blockchain:</span>
        <div className="flex px-2 space-x-2">
          <a
            target="blank"
            className="text-blue-400"
            href={props.network === Network.MAINNET ? "https://blockstream.info/" : "https://blockstream.info/testnet/"}
          >
            Bitcoin
          </a>
          <a target="blank" className="text-blue-400" href="https://wallet.bitshares.org/#/">
            Bitshares
          </a>
        </div>
      </div>
    </div>
  )
}
