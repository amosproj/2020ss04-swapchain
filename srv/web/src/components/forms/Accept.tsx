import React, { useState, useEffect } from "react"
import { fakeKey } from "../../util"
import { Label } from "./Label"
import { Input } from "./Input"
import { SubmitButton } from "./SubmitButton"
import { Form } from "./Form"
import { ReactComponent as Exclamation } from "../../icons/exclamation.svg"
import { Spinner } from "../Spinner"
import { RadioButton } from "./RadioButton"
import { State, Network, Timelock, Priority } from "./enums"

export const Accept = () => {
    // Application stae for handling the flow
    const [state, setState] = useState(State.IDLE)
    const [isValid, setValid] = useState(false)

    // Used to display helpful error messages to the user
    const [errorMessage, setErrorMessage] = useState("")

    // Collects the user input in one place
    const [fields, setFields] = useState({
        networkToTrade: Network.MAINNET,
        hash: "",
        bitcoinPrivateKey: "",
        bitcoinPublicKey: "",
        bitsharesPrivateKey: "",
        bitsharesAccountName: "",
        counterpartyBitcoinPublicKey: "",
        counterpartyBitsharesAccountName: "",
        timelockDuration: Timelock.SHORT,
        orderDuration: 7,
        priority: Priority.HIGH,
    })

    /**
     * Validate all fields.
     * @returns The respective error messsage or empty string if valid.
     */
    const validate = (): string => {
        if (fields.hash === "") {
            return "Hash is empty"
        }
        if (fields.bitcoinPrivateKey.length === 32) {
            return "Bitcoin private key is not 32 characters long"
        }
        if (fields.bitsharesPrivateKey === "") {
            return "Bitshares private key is empty"
        }
        if (fields.counterpartyBitcoinPublicKey === "") {
            return "Counterparty bitcoin public key is empty"
        }
        if (fields.counterpartyBitsharesAccountName === "") {
            return "Counterparty bitshares account name is empty"
        }
        return ""
    }

    // Go back to idle when the user fixes their input errors.
    // Does nothing if the accs is already running
    useEffect(() => {
        // Failure state is final
        if (state !== State.FAILURE) {
            if (state !== State.RUNNING && isValid) {
                setErrorMessage("")
                setState(State.IDLE)
            }
        }
    }, [fields, state, isValid])


    /**
     * Update the field state from a FormEvent
     * @param e - HTMLInput FormEvent.
     */
    const updateField = (e: React.FormEvent<HTMLInputElement>) => {
        setFields({
            ...fields,
            [e.currentTarget.name]: e.currentTarget.value
        })
    }

    /**
    * Update the field state manually via onClick for example.
    * @param key - The key in the field state.
    * @param value - New value to be stored.
    */
    const updateFieldByKey = (key: string, value: number | string) => {
        setFields({
            ...fields,
            [key]: value
        })
    }

    /**
     * This hook handles the SubmitButton.
     * When a user thinks he is done filling out the form it is validated first and feedback is given to the user.
     */
    const submitHandler = () => {
        const errorMessage = validate()
        setValid(errorMessage === "")
        
        // Checking "manually" because the state is updated asynchronously
        if (errorMessage !== "") {
            setErrorMessage(errorMessage)
            setState(State.INVALID_INPUT)
            return
        }


        setState(State.RUNNING)

        // TODO: Replace with accs
        setTimeout(() => {
            if (Math.random() > 0.5) {
                setState(State.SUCCESS)
            } else {
                setState(State.ERROR)
                setErrorMessage("Joab didn't do his job, should we fire him?")
            }
        }, 5000)
    }

    return (
        <Form
            main={
                <div className="flex flex-col justify-center w-full p-8">
                    <h2 className="text-3xl font-semibold leading-tight text-gray-800">Accept an Atomic Cross Chain Swap</h2>
                    <div className="flex flex-col mt-12 space-y-10">
                        <section>
                            <Label label="On what network do you want to trade"></Label>
                            <div className="flex flex-col justify-center space-y-4 md:flex-row md:space-x-4 md:space-y-0">
                                <RadioButton
                                    description="You are sending real money!"
                                    name="Mainnet"
                                    tag={<Exclamation className={`h-8 text-red-600 ${fields.networkToTrade === Network.MAINNET ? "" : "hidden"}`}></Exclamation>}
                                    onClick={() => updateFieldByKey("networkToTrade", Network.MAINNET)}
                                    selected={fields.networkToTrade === Network.MAINNET}
                                ></RadioButton>
                                <RadioButton
                                    description="Send test currencies on the testnet"
                                    name="Testnet"
                                    onClick={() => updateFieldByKey("networkToTrade", Network.TESTNET)}
                                    selected={fields.networkToTrade === Network.TESTNET}
                                ></RadioButton>
                            </div>
                        </section>
                        <section>
                            <h2 className="text-xl font-semibold leading-tight text-gray-800">Your Data</h2>
                            <div className="items-center justify-between mt-4 md:space-x-4 md:flex">

                                <div className="md:w-1/2">
                                    <Label label="Bitcoin private key"></Label>
                                    <Input
                                        name="bitcoinPrivateKey"
                                        onChange={updateField}
                                        placeholder={fakeKey(30, fields.networkToTrade)}
                                        type="text"
                                        value={fields.bitcoinPrivateKey}
                                    ></Input>
                                </div>
                                <div className="md:w-1/2">
                                    <Label label="Bitshares private key"></Label>
                                    <Input
                                        name="bitsharesPrivateKey"
                                        onChange={updateField}
                                        placeholder="amos"
                                        type="text"
                                        value={fields.bitsharesPrivateKey}
                                    ></Input>
                                </div>
                            </div>
                            <p className="px-4 mx-auto mt-2 text-sm text-center text-gray-500">
                                Your public Bitcoin key and Bitshares account name can be calculated from your private key. Your private keys will never leave your browser, it is only used to sign your transactions. <a href="/" className="relative text-xs text-blue-500">Read more in our docs.</a>
                            </p>
                        </section>

                        <section>
                            <h2 className="text-xl font-semibold leading-tight text-gray-800">Counterparty Data</h2>
                            <div className="items-center justify-between mt-4 md:space-x-4 md:flex">

                                <div className="md:w-1/2">
                                    <Label label="Bitcoin public key"></Label>
                                    <Input
                                        name="counterpartyBitcoinPublicKey"
                                        onChange={updateField}
                                        placeholder={fakeKey(30, fields.networkToTrade)}
                                        type="text"
                                        value={fields.counterpartyBitcoinPublicKey}
                                    ></Input>
                                </div>
                                <div className="md:w-1/2">
                                    <Label label="Bitshares account name"></Label>
                                    <Input
                                        name="counterpartyBitsharesAccountName"
                                        onChange={updateField}
                                        placeholder=""
                                        type="text"
                                        value={fields.counterpartyBitsharesAccountName}
                                    ></Input>
                                </div>
                            </div>
                        </section>

                        <section>
                            <div className="block">
                                <Label label="Hash from Counterparty"></Label>
                                <Input name="hash" value={fields.hash} onChange={updateField} type="text" placeholder="Please enter the hash you received from your trading partner"></Input>
                            </div>

                        </section>
                        <section>
                            <h2 className="text-xl font-semibold leading-tight text-gray-800">HTLC settings</h2>
                            <div className="mt-4">
                                <Label label="Choose your timelock"></Label>
                                <div className="flex flex-col space-y-4 md:flex-row md:space-x-4 md:space-y-0">
                                    <RadioButton
                                        description="Around 1 hour. This is the shortest duration possible while making sure the transactions are confirmed in time."
                                        hint="6 blocks"
                                        name={Timelock[Timelock.SHORT]}
                                        onClick={() => updateFieldByKey("timelockDuration", Timelock.SHORT)}
                                        selected={fields.timelockDuration === Timelock.SHORT}
                                    ></RadioButton>
                                    <RadioButton
                                        description="Around 2 hours. Offers more time for the counterparty to come online."
                                        hint="13 blocks"
                                        name={Timelock[Timelock.MEDIUM]}
                                        onClick={() => updateFieldByKey("timelockDuration", Timelock.MEDIUM)}
                                        selected={fields.timelockDuration === Timelock.MEDIUM}
                                    ></RadioButton>
                                    <RadioButton
                                        description="Around 3 hours. Gives your counterparty even more time."
                                        hint="20 blocks"
                                        name={Timelock[Timelock.LONG]}
                                        onClick={() => updateFieldByKey("timelockDuration", Timelock.LONG)}
                                        selected={fields.timelockDuration === Timelock.LONG}
                                    ></RadioButton>
                                </div>
                                <p className="px-4 mx-auto mt-2 text-sm text-center text-gray-500">
                                    By design one block should equal 10 minutes. However it can differ quite a lot in reality. Please make sure you enter the same value as your trading partner.
                                    </p>
                            </div>
                        </section>

                        <section>
                            <Label label="Choose your priority"></Label>
                            <div className="flex flex-col space-y-4 md:flex-row md:space-x-4 md:space-y-0">
                                <RadioButton
                                    description="You pay the highest fees to increase the chance for your transaction to be picked up by the miners."
                                    name={Priority[Priority.HIGH]}
                                    onClick={() => updateFieldByKey("priority", Priority.HIGH)}
                                    selected={fields.priority === Priority.HIGH}
                                ></RadioButton>
                                <RadioButton
                                    description="You pay a moderate amount of fees so miners will probably confirm your transaction soon."
                                    name={Priority[Priority.MEDIUM]}
                                    onClick={() => updateFieldByKey("priority", Priority.MEDIUM)}
                                    selected={fields.priority === Priority.MEDIUM}
                                ></RadioButton>
                                <RadioButton
                                    description="You pay the lowest fees but might have to wait a few more blocks for your transaction to be confirmed."
                                    name={Priority[Priority.LOW]}
                                    onClick={() => updateFieldByKey("priority", Priority.LOW)}
                                    selected={fields.priority === Priority.LOW}
                                ></RadioButton>
                            </div>
                        </section>
                    </div>
                </div>
            }
            footer={<>
                {(function () {
                    switch (state) {
                        case State.IDLE:
                            return <SubmitButton borderColor="teal" label="Submit" onClick={submitHandler}></SubmitButton>
                        case State.INVALID_INPUT:
                            return < div className="flex flex-col">
                                <p className="pb-4 text-red-400 text-bold">{errorMessage}</p>
                                <SubmitButton borderColor="gray" label="Submit" onClick={submitHandler}></SubmitButton>
                            </div>
                        case State.RUNNING:
                            return <Spinner className="h-20 bg-transparent"></Spinner>
                        case State.SUCCESS:
                            return <span className="p-10 text-teal-400 text-bold">SUCCESS</span>
                        case State.ERROR:
                            return < div className="flex flex-col">
                                <p className="pb-4 text-red-400 text-bold">{errorMessage}</p>
                                <SubmitButton borderColor="gray" label="Try again" onClick={submitHandler}></SubmitButton>
                            </div>
                        case State.FAILURE:
                            return <p className="pb-4 text-xl text-red-400 text-bold">{errorMessage}</p>
                           
                    }
                })()}
            </>
            }></Form>
    )
}