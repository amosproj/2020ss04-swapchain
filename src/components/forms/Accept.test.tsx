import React from "react"
import { Accept } from "./Accept"
import renderer from "react-test-renderer"
import * as util from "../../util"
jest.mock("../../util")

jest.spyOn(util, "fakeKey").mockReturnValue("randomLookingHash")

it("renders correctly", () => {
  const tree = renderer.create(<Accept></Accept>).toJSON()
  expect(tree).toMatchSnapshot()
})
