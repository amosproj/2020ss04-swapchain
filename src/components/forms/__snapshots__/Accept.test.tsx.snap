// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="flex flex-col items-center w-full bg-white border rounded shadow-lg"
>
  <div
    className="flex flex-col justify-center w-full p-4 space-y-4"
  >
    <div
      className="flex flex-col justify-center w-full p-8"
    >
      <h2
        className="text-3xl font-semibold leading-tight text-gray-800"
      >
        Accept an Atomic Cross Chain Swap
      </h2>
      <div
        className="flex flex-col mt-12 space-y-10"
      >
        <section>
          <label
            className="text-xs leading-4 tracking-wider text-gray-700 uppercase"
          >
            On what network do you want to trade
          </label>
          <div
            className="flex flex-col justify-center space-y-4 md:flex-row md:space-x-4 md:space-y-0"
          >
            <button
              className="focus:outline-none flex w-full items-center justify-between border rounded border-teal-400 shadow hover:bg-white hover:shadow-sm"
              onClick={[Function]}
            >
              <div
                className="flex flex-col items-start justify-between p-4"
              >
                <div
                  className="flex items-center justify-between w-full"
                >
                  <span
                    className="text-left font-semibold text-gray-900"
                  >
                    Mainnet
                  </span>
                  <span
                    className="text-xs font-thin text-gray-600"
                  />
                </div>
                <p
                  className="mt-2 text-sm text-left text-gray-600"
                >
                  You are sending real money!
                </p>
              </div>
              <div
                className="flex items-center justify-center"
              >
                <span
                  className="mr-4 text-3xl font-semibold hidden md:block text-gray-900 "
                >
                  <svg
                    className="h-8 text-red-600 "
                  >
                    exclamation.svg
                  </svg>
                </span>
              </div>
            </button>
            <button
              className="focus:outline-none flex w-full items-center justify-between border rounded bg-gray-100 hover:bg-white hover:shadow-sm"
              onClick={[Function]}
            >
              <div
                className="flex flex-col items-start justify-between p-4"
              >
                <div
                  className="flex items-center justify-between w-full"
                >
                  <span
                    className="text-left font-semibold text-gray-600"
                  >
                    Testnet
                  </span>
                  <span
                    className="text-xs font-thin text-gray-400"
                  />
                </div>
                <p
                  className="mt-2 text-sm text-left text-gray-400"
                >
                  Send test currencies on the testnet
                </p>
              </div>
              <div
                className="flex items-center justify-center"
              >
                <span
                  className="mr-4 text-3xl font-semibold hidden md:block text-gray-300"
                />
              </div>
            </button>
          </div>
        </section>
        <section>
          <h2
            className="text-xl font-semibold leading-tight text-gray-800"
          >
            Your Data
          </h2>
          <div
            className="items-center justify-between mt-4 md:space-x-4 md:flex"
          >
            <div
              className="md:w-1/2"
            >
              <label
                className="text-xs leading-4 tracking-wider text-gray-700 uppercase"
              >
                Bitcoin private key
              </label>
              <input
                className="w-full py-3 font-mono text-center text-gray-700 bg-gray-100 border border-gray-200 rounded focus:bg-white focus:border-teal-400"
                name="bitcoinPrivateKey"
                onChange={[Function]}
                placeholder="randomLookingHash"
                required={true}
                type="text"
                value=""
              />
            </div>
            <div
              className="md:w-1/2"
            >
              <label
                className="text-xs leading-4 tracking-wider text-gray-700 uppercase"
              >
                Bitshares private key
              </label>
              <input
                className="w-full py-3 font-mono text-center text-gray-700 bg-gray-100 border border-gray-200 rounded focus:bg-white focus:border-teal-400"
                name="bitsharesPrivateKey"
                onChange={[Function]}
                placeholder="amos"
                required={true}
                type="text"
                value=""
              />
            </div>
          </div>
          <p
            className="px-4 mx-auto mt-2 text-sm text-center text-gray-500"
          >
            Your public Bitcoin key and Bitshares account name can be calculated from your private key. Your private keys will never leave your browser, it is only used to sign your transactions.
             
            <a
              className="relative text-xs text-blue-500"
              href="/"
            >
              Read more in our docs.
            </a>
          </p>
        </section>
        <section>
          <h2
            className="text-xl font-semibold leading-tight text-gray-800"
          >
            Counterparty Data
          </h2>
          <div
            className="items-center justify-between mt-4 md:space-x-4 md:flex"
          >
            <div
              className="md:w-1/2"
            >
              <label
                className="text-xs leading-4 tracking-wider text-gray-700 uppercase"
              >
                Bitcoin public key
              </label>
              <input
                className="w-full py-3 font-mono text-center text-gray-700 bg-gray-100 border border-gray-200 rounded focus:bg-white focus:border-teal-400"
                name="counterpartyBitcoinPublicKey"
                onChange={[Function]}
                placeholder="randomLookingHash"
                required={true}
                type="text"
                value=""
              />
            </div>
            <div
              className="md:w-1/2"
            >
              <label
                className="text-xs leading-4 tracking-wider text-gray-700 uppercase"
              >
                Bitshares account name
              </label>
              <input
                className="w-full py-3 font-mono text-center text-gray-700 bg-gray-100 border border-gray-200 rounded focus:bg-white focus:border-teal-400"
                name="counterpartyBitsharesAccountName"
                onChange={[Function]}
                placeholder=""
                required={true}
                type="text"
                value=""
              />
            </div>
          </div>
        </section>
        <section>
          <div
            className="block"
          >
            <label
              className="text-xs leading-4 tracking-wider text-gray-700 uppercase"
            >
              Hash from Counterparty
            </label>
            <input
              className="w-full py-3 font-mono text-center text-gray-700 bg-gray-100 border border-gray-200 rounded focus:bg-white focus:border-teal-400"
              name="hash"
              onChange={[Function]}
              placeholder="Please enter the hash you received from your trading partner"
              required={true}
              type="text"
              value=""
            />
          </div>
        </section>
        <section>
          <h2
            className="text-xl font-semibold leading-tight text-gray-800"
          >
            HTLC settings
          </h2>
          <div
            className="mt-4"
          >
            <label
              className="text-xs leading-4 tracking-wider text-gray-700 uppercase"
            >
              Choose your timelock
            </label>
            <div
              className="flex flex-col space-y-4 md:flex-row md:space-x-4 md:space-y-0"
            >
              <button
                className="focus:outline-none flex w-full items-center justify-between border rounded border-teal-400 shadow hover:bg-white hover:shadow-sm"
                onClick={[Function]}
              >
                <div
                  className="flex flex-col items-start justify-between p-4"
                >
                  <div
                    className="flex items-center justify-between w-full"
                  >
                    <span
                      className="text-left font-semibold text-gray-900"
                    >
                      SHORT
                    </span>
                    <span
                      className="text-xs font-thin text-gray-600"
                    >
                      6 blocks
                    </span>
                  </div>
                  <p
                    className="mt-2 text-sm text-left text-gray-600"
                  >
                    Around 1 hour. This is the shortest duration possible while making sure the transactions are confirmed in time.
                  </p>
                </div>
                <div
                  className="flex items-center justify-center"
                >
                  <span
                    className="mr-4 text-3xl font-semibold hidden md:block text-gray-900 "
                  />
                </div>
              </button>
              <button
                className="focus:outline-none flex w-full items-center justify-between border rounded bg-gray-100 hover:bg-white hover:shadow-sm"
                onClick={[Function]}
              >
                <div
                  className="flex flex-col items-start justify-between p-4"
                >
                  <div
                    className="flex items-center justify-between w-full"
                  >
                    <span
                      className="text-left font-semibold text-gray-600"
                    >
                      MEDIUM
                    </span>
                    <span
                      className="text-xs font-thin text-gray-400"
                    >
                      13 blocks
                    </span>
                  </div>
                  <p
                    className="mt-2 text-sm text-left text-gray-400"
                  >
                    Around 2 hours. Offers more time for the counterparty to come online.
                  </p>
                </div>
                <div
                  className="flex items-center justify-center"
                >
                  <span
                    className="mr-4 text-3xl font-semibold hidden md:block text-gray-300"
                  />
                </div>
              </button>
              <button
                className="focus:outline-none flex w-full items-center justify-between border rounded bg-gray-100 hover:bg-white hover:shadow-sm"
                onClick={[Function]}
              >
                <div
                  className="flex flex-col items-start justify-between p-4"
                >
                  <div
                    className="flex items-center justify-between w-full"
                  >
                    <span
                      className="text-left font-semibold text-gray-600"
                    >
                      LONG
                    </span>
                    <span
                      className="text-xs font-thin text-gray-400"
                    >
                      20 blocks
                    </span>
                  </div>
                  <p
                    className="mt-2 text-sm text-left text-gray-400"
                  >
                    Around 3 hours. Gives your counterparty even more time.
                  </p>
                </div>
                <div
                  className="flex items-center justify-center"
                >
                  <span
                    className="mr-4 text-3xl font-semibold hidden md:block text-gray-300"
                  />
                </div>
              </button>
            </div>
            <p
              className="px-4 mx-auto mt-2 text-sm text-center text-gray-500"
            >
              By design one block should equal 10 minutes. However it can differ quite a lot in reality. Please make sure you enter the same value as your trading partner.
            </p>
          </div>
        </section>
        <section>
          <label
            className="text-xs leading-4 tracking-wider text-gray-700 uppercase"
          >
            Choose your priority
          </label>
          <div
            className="flex flex-col space-y-4 md:flex-row md:space-x-4 md:space-y-0"
          >
            <button
              className="focus:outline-none flex w-full items-center justify-between border rounded border-teal-400 shadow hover:bg-white hover:shadow-sm"
              onClick={[Function]}
            >
              <div
                className="flex flex-col items-start justify-between p-4"
              >
                <div
                  className="flex items-center justify-between w-full"
                >
                  <span
                    className="text-left font-semibold text-gray-900"
                  >
                    HIGH
                  </span>
                  <span
                    className="text-xs font-thin text-gray-600"
                  />
                </div>
                <p
                  className="mt-2 text-sm text-left text-gray-600"
                >
                  You pay the highest fees to increase the chance for your transaction to be picked up by the miners.
                </p>
              </div>
              <div
                className="flex items-center justify-center"
              >
                <span
                  className="mr-4 text-3xl font-semibold hidden md:block text-gray-900 "
                />
              </div>
            </button>
            <button
              className="focus:outline-none flex w-full items-center justify-between border rounded bg-gray-100 hover:bg-white hover:shadow-sm"
              onClick={[Function]}
            >
              <div
                className="flex flex-col items-start justify-between p-4"
              >
                <div
                  className="flex items-center justify-between w-full"
                >
                  <span
                    className="text-left font-semibold text-gray-600"
                  >
                    MEDIUM
                  </span>
                  <span
                    className="text-xs font-thin text-gray-400"
                  />
                </div>
                <p
                  className="mt-2 text-sm text-left text-gray-400"
                >
                  You pay a moderate amount of fees so miners will probably confirm your transaction soon.
                </p>
              </div>
              <div
                className="flex items-center justify-center"
              >
                <span
                  className="mr-4 text-3xl font-semibold hidden md:block text-gray-300"
                />
              </div>
            </button>
            <button
              className="focus:outline-none flex w-full items-center justify-between border rounded bg-gray-100 hover:bg-white hover:shadow-sm"
              onClick={[Function]}
            >
              <div
                className="flex flex-col items-start justify-between p-4"
              >
                <div
                  className="flex items-center justify-between w-full"
                >
                  <span
                    className="text-left font-semibold text-gray-600"
                  >
                    LOW
                  </span>
                  <span
                    className="text-xs font-thin text-gray-400"
                  />
                </div>
                <p
                  className="mt-2 text-sm text-left text-gray-400"
                >
                  You pay the lowest fees but might have to wait a few more blocks for your transaction to be confirmed.
                </p>
              </div>
              <div
                className="flex items-center justify-center"
              >
                <span
                  className="mr-4 text-3xl font-semibold hidden md:block text-gray-300"
                />
              </div>
            </button>
          </div>
        </section>
      </div>
    </div>
  </div>
  <footer
    className="flex justify-center w-full p-4 bg-gray-100 border-t border-gray-200"
  >
    <button
      className="px-8 py-3 font-semibold text-gray-900 uppercase transition duration-200 ease-in-out bg-white border rounded-sm shadow-sm focus:outline-none hover:border-teal-500 hover:bg-white border-teal-300"
      onClick={[Function]}
      type="button"
    >
      Submit
    </button>
  </footer>
</div>
`;
